components/scheduler/executors.go

```diff
@@ -23,41 +23,472 @@
+	}
+
+	res, err := e.options.FrontendClient.StartWorkflowExecution(context.Background(), req)
+	if err != nil {
+		// TODO(Tianyu): original implementation translates this error which may not be relevant any more
+		// TODO(Tianyu): On failure, need to figure out how HSM-based implementation should retry, as we can no longer rely on automatic activity retries.
+		return nil, err
+	}
+	// this will not match the time in the workflow execution started event
+	// exactly, but it's just informational so it's close enough.
+	now := time.Now()
+
+	if !start.Manual {
+		// record metric only for _scheduled_ actions, not trigger/backfill, otherwise it's not meaningful
+		desiredTime := cmp.Or(start.DesiredTime, start.ActualTime)
+		e.options.MetricsHandler.Timer(metrics.ScheduleActionDelay.Name()).Record(now.Sub(desiredTime.AsTime()))
+	}
+
+	return &schedpb.ScheduleActionResult{
+		ScheduleTime: start.ActualTime,
+		ActualTime:   timestamppb.New(now),
+		StartWorkflowResult: &commonpb.WorkflowExecution{
+			WorkflowId: workflowID,
+			RunId:      res.RunId,
+		},
+	}, nil
+}
+
+func (e taskExecutor) terminateWorkflow(s *Scheduler, ex *commonpb.WorkflowExecution) {
+	// TODO: remove after https://github.com/temporalio/sdk-go/issues/1066
+	// TODO(Tianyu): hardcoded wait time
+	workflowCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	rreq := &historyservice.TerminateWorkflowExecutionRequest{
+		NamespaceId: s.Args.State.NamespaceId,
+		TerminateRequest: &workflowservice.TerminateWorkflowExecutionRequest{
+			Namespace: s.Args.State.Namespace,
+			// only set WorkflowId so we cancel the latest, but restricted by FirstExecutionRunId
+			WorkflowExecution:   &commonpb.WorkflowExecution{WorkflowId: ex.WorkflowId},
+			Reason:              "terminated by schedule overlap policy",
+			Identity:            s.identity(),
+			FirstExecutionRunId: ex.RunId,
+		},
+	}
+	_, err := e.options.HistoryClient.TerminateWorkflowExecution(workflowCtx, rreq)
+
+	// TODO(Tianyu): original implementation translates this error which may not be relevant any more
+	// TODO(Tianyu): On failure, need to figure out how HSM-based implementation should retry, as we can no longer rely on automatic activity retries.
+	if err != nil {
+		e.options.Logger.Error("terminate workflow failed", tag.WorkflowID(ex.WorkflowId), tag.Error(err))
+		e.options.MetricsHandler.Counter(metrics.ScheduleTerminateWorkflowErrors.Name()).Record(1)
+	}
+}
+
+func (e taskExecutor) cancelWorkflow(s *Scheduler, ex *commonpb.WorkflowExecution) {
+	// TODO: remove after https://github.com/temporalio/sdk-go/issues/1066
+	// TODO(Tianyu): hardcoded wait time
+	workflowCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	rreq := &historyservice.RequestCancelWorkflowExecutionRequest{
+		NamespaceId: s.Args.State.NamespaceId,
+		CancelRequest: &workflowservice.RequestCancelWorkflowExecutionRequest{
+			Namespace: s.Args.State.Namespace,
```

Comment:
You can load the namespace from the registry and get the "current" name in case the namespace was renamed since this scheduler was created.

Reply:
Added an update step per invocation of state machine logic. The rest of this code path has been converted to use namespace id wherever possible.